[1단계 - 자동차 경주] 니야(최수연) 미션 제출합니다.

sooyeoniya opened the pull request:
# 🐥 니야와 호이초이의 페어 프로그래밍 🐥
### 안녕하세요, 니야(최수연) 1단계 자동차 경주 미션 제출하겠습니다!

# 🛠️ 설계 방식

### ✔️ 개별 Car 클래스 생성 및 관리

프리코스 때는 우리 둘 다 Cars 라는 클래스에서 모든 자동차의 정보를 한 번에 관리했었다.
근데 이번에는 각 자동차별로 name과 count 값을 관리하는 각각의 인스턴스를 생성해서 관리하도록 진행했다.
그러다보니 외부에서 전체 자동차에 대한 인스턴스를 관리해주어야 했다.

### ✔️ Controller 분리 (GameController와 RaceController)

처음에는 입출력과 유효성 검증 로직과 전체 경주를 실행하고 최종 우승자를 추출하는 로직이 모두 하나의 Controller에서 관리되고 있었다.
이렇게 하나의 Controller에서 관리하다보니 코드가 길어지고 기능이 많아져서 응집도가 높아졌다.
그래서 자동차 경주를 실행하는 로직 자체를 분리하고자 하였다. 결론적으로 아래와 같이 분류하게 되었다.

- `GameController`: 게임 자체에 대한 전체 실행 로직 관리
- `RaceController`: 시도횟수에 따른 각각의 경주를 실행하고 우승자를 추출하는 로직 관리

### ✔️ 입출력 로직 Console 객체로 관리

처음에는 console.log()와 readLineAsync()를 별도로 사용하고 있었다.
그러나 직관적으로 입출력 로직임을 인지하기 어려웠고, 사용 방식에도 일관성이 부족하다고 느꼈다.
이에 따라 입출력 역할만을 담당하는 Console 객체를 만들어 메서드로 관리하면,
우테코에서 제공해준 라이브러리처럼 명확하게 '입출력 책임을 분리했다'는 느낌을 줄 수 있겠다고 생각했다.
이후 모든 입출력 처리를 Console.print()와 Console.readLineAsync() 등 메서드로 일관되게 사용하도록 변경했다.

# 🧩 실행 로직 도식화

```
index.js -> GameController 인스턴스 생성 후, play() (게임 시작)
.
├── GameController
├── readAndValidateInputs (입력값 받기)
├── RaceController (입력값(자동차들, 시도횟수) 바탕으로 경주)
│  ├── race (경주 시작)
│  │   └── round (라운드 별 진행)
│  └── getWinners (우승자 선별)
└── printWinner (우승자 출력)
```
# ❓궁금한 점
### ✔️ 클래스의 정적 메서드와 객체의 메서드 차이
프리코스에서 제공해준 missonUtils의 Console 모듈은 클래스기반의 정적메서드로 구현이 되어 있었습니다.
하지만 저희는 Input, Output 모듈을 객체로 관리해서 개발을 진행하고 있었습니다.
리팩토링 과정에서 Console 모듈을 참고하면서 클래스의 정적 메서드로 관리하는 방법과 객체의 메서드로 관리하는 방식의 차이에 대해 궁금해졌습니다.
두 방식의 차이점에 대해 알아봤는데, 정확히 기준을 잘 모르겠습니다.  help me.

sooyeoniya committed:
docs: README.md 기능 요구사항 정리

sooyeoniya committed:
docs: README.md 페어 규칙 및 회고 템플릿 작성

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
docs: 예외사항 추가
dquote>
dquote> Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
style: prettier 적용

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
feat(Input): 입력 로직 구현

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
feat(validation): 자동차 유효성 검증 구현 및 prettier 적용

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
docs: 25.02.12 KPT 회고 작성

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
feat(validation): 시도 횟수 유효성 검증 구현

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
feat(utils): 무작위 값 생성 함수 구현

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
feat(Car): 자동차 클래스 생성

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
feat(Output): 시도 횟수 별 자동차 경주 내역 출력

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
feat(Winners): 최종 우승자 추출 및 출력

Co-authored-by: hoyyChoi <hoyeon8165@gmail.com>

sooyeoniya committed:
docs: 공동기여자 이메일 설정 방법 수정

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
refactor(Console): 입출력 객체 분리

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
refactor: 입출력 로직에 맞춰 전체 파일 리팩토링

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
docs: README 내용 분리

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
chore: 주석 및 줄바꿈 제거

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
style: prettier printWidth 200으로 변경

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
refactor: 함수길이 및 indent depth 수정

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
refactor: 메시지 및 기준값 상수 처리

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
refactor: Controller를 Game과 Race로 분리

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
test: Car 클래스 테스트 구현

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
test: validateCarNames 함수 테스트 구현

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
test: validateTryCount 함수 테스트 구현

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
test: RaceController 클래스 테스트 구현

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
docs: 데일리 회고 작성

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
chore: validations 폴더 이름 수정

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

sooyeoniya committed:
docs: 기능 명세서 수정

Co-authored-by: hoyyChoi <hoyyChoi@users.noreply.github.com>

Creative-Lee commented on README.md:
```diff
@@ -1 +1,127 @@
+- [x] 자동차 이름이 중복되는 경우
+
+**시도횟수 입력에 대한 유효성 검사**
+
+- [x] 자연수(양의 정수)가 아닌 경우
+- [x] 0이하인 경우
+
```
✔️ 
악성 사용자 '크론'이 100000000000000000000000000000000000000000000회를 시도합니다 😈 
![Feb-14-2025 20-43-41](https://github.com/user-attachments/assets/425c8e73-8b3c-40dc-94b6-99b1ba8d2fa1)
게임이 끝나지 않아요. 예외처리가 필요하겠네요~!

sooyeoniya replied:
페어랑 최댓값 설정하는 부분에서 고민을 했는데, 시도횟수 최댓값을 얼마로 설정해야 할지 고민이 되어서 나중으로 미루다가 결국 처리하지 못하고 냈던 것 같네용! 지금 한 번 설정해보겠습니다 :)

Creative-Lee commented on README.md:
```diff
@@ -1 +1,127 @@
+
+### ✅ 예외 사항
+
+**자동차 입력에 대한 유효성 검사**
+
+- [x] 자동차 이름 1자 이상 ~ 5자 이하
+- [x] 자동차 이름 전체가 공백인 경우
+- [x] 자동차 이름이 중복되는 경우
+
```
✔️ 
혼자서 게임을 실행할 수 있네요.
![크론 혼자](https://github.com/user-attachments/assets/0be29839-7cfe-4b39-bb4b-ea351e13f1f9)

크론이 외롭지 않을까요? 😢  예외처리를 해줍시다!

sooyeoniya replied:
크론 혼자 경주를 진행하는 경우를 고려하지 못했군요..! 이 부분은 프리코스 때도 생각을 아예 하지 못했던 부분인데, 지적 감사합니다 ! 👍🏻

Creative-Lee commented on README.md:
```diff
@@ -1 +1,127 @@
+- [x] 자동차 경주 내역 출력
+- [x] 최종 우승자 출력
+
+### ✅ 리팩토링
+
+- [x] 입출력 따로 관리 (Console 객체)
+- [x] 함수 길이 15 제한
+- [x] indent depth 2 제한
+- [x] 상수 처리
+- [x] Controller 분리 (Game, Race)
```
리팩토링도 체크하셨다니 훌륭하네요~! 👍

Creative-Lee commented on docs/retrospect.md:
```diff
@@ -0,0 +1,87 @@
+
+- jest 모킹 함수 더 공부하기.
+- 정규표현식 더 공부하기.
+- 변수명 잘 짓는 연습하기. (영어 이슈;;)
```
유 윌 겟 베러 !
  
  
  
  
  
  
(영어 좀 잘해보이나요?)

sooyeoniya replied:
도밥은 영어 잘하실 것 같아용 😁

Creative-Lee commented on src/constants/constants.js:
```diff
@@ -0,0 +1,25 @@
+export const RANDOM_BOUNDARY_NUMBER = 4;
```
어떤 임계값을 나타낼때 threshold 라는 단어도 많이 쓴답니다~!

sooyeoniya replied:
헉 맞네요, 훨씬 좋은 단어인 것 같습니다 ㅎㅎ 감사합니다 !

Creative-Lee commented on src/constants/constants.js:
```diff
@@ -0,0 +1,25 @@
+  TRY_COUNT: "시도할 횟수는 몇 회인가요?\n",
+});
+
+export const OUTPUT_PROMPT_MESSAGE = Object.freeze({
+  RACE_RESULT: "\n실행 결과",
```
`Object.freeze` 사용은 따봉을 2개 드리고 싶네요 👍  👍 

여유되시면 `Object.freeze`의 동작에는 어떤 한계가 있는지도 학습해보시면 좋겠습니다 ㅎㅎ

sooyeoniya replied:
`Object.freeze` 한계에서 제가 알고 있는 건, 깊은 동결이 안된다는 문제점인데요. 이 때문에 프리코스 당시 리뷰를 받아서 깊은 동결하는 방법을 학습하여 적용한 적이 있습니다. 이번 페어 프로그래밍 때도 `ERROR_MESSAGE` 를 중첩된 객체로 만들어 한 번에 묶으려다가 깊은 동결을 위한 함수를 따로 구현해야해서 (약간의 번거로움 + 우선순위에서 밀림 때문에) `carName`과 `tryCount` 두 개의 에러 메시지를 따로 분리했던 것도 있습니다. 혹시 이 문제 말고 `Object.freeze`의 다른 한계가 있다면 공유해주시면 감사하겠습니다 !

Creative-Lee replied:
이미 알고 계시다니 훌륭합니다~! 👍

Creative-Lee commented on src/constants/constants.js:
```diff
@@ -0,0 +1,25 @@
+});
+
+export const ERROR_CAR_NAMES_MESSAGE = Object.freeze({
+  NOT_EXIST: "자동차 이름이 존재하지 않습니다.",
+  OVER: "자동차 이름이 5글자가 넘습니다.",
```
에러메세지는 세상 친절할수록 좋다고 생각해요. 

자동차 이름이 존재하지 않습니다. 의 느낌과
자동차 이름이 5글자가 넘습니다. 는 조금 늬양스가 다른것 같지않나요? 

비슷하게 작성해보면 자동차 이름은 5글자 이하여야 합니다. 정도가 되겠네요~!

sooyeoniya replied:
하나하나 꼼꼼하게 살펴볼 줄 아는 도밥을 보고 배워야겠다는 생각이 드네요.. 사소한 디테일까지 신경써주셔서 감사합니다 :)

Creative-Lee commented on src/constants/constants.js:
```diff
@@ -0,0 +1,25 @@
+export const ERROR_TRY_COUNT_MESSAGE = Object.freeze({
+  INVALID_NUMBER: "시도 횟수가 숫자가 아닙니다.",
+  INVAILD_INTEGER: "시도 횟수가 정수가 아닙니다.",
+  INVALID_RANGE: "시도 횟수는 0 이하가 될 수 없습니다.",
```
요기 메세지들도 위 코멘트와 비슷한 맥락으로 개선해볼 수 있겠죠?

Creative-Lee commented on src/controllers/RaceController.js:
```diff
@@ -0,0 +1,38 @@
+import Output from "../views/Output.js";
+
+class RaceController {
+  #carsInstance = [];
```
필드명에서 `Instance` 라는 단어가 빠져도 어색하지 않게 느껴져요.
배열을 초기화 할때 new Car()만 보고서도 Instance라는 건 알수 있고요.
작명에 고민했던 포인트가 있었을까요?

sooyeoniya replied:
처음에는 `cars`라고 하려 했는데, 각 자동차를 모아둔 인스턴스 배열이라는 것을 보여주고 싶어서 저런식으로 작명한 것 같습니다. `RaceController` 내부에 `Car`에 대한 생성자 함수가 있으니 굳이 다시 `Instance`라는 것을 명시할 필요는 없겠네요! 😃

Creative-Lee commented on docs/pair-rule.md:
```diff
@@ -0,0 +1,15 @@
+
+## ➡️ **내비게이터**
+
+- 쪼지 말 것.
```
ㅋㅋㅋㅋㅋㅋㅋㅋㅋ 🐦

Creative-Lee commented on src/models/Car.js:
```diff
@@ -0,0 +1,23 @@
+    this.#name = name;
+  }
+
+  move(randomNumber) {
```
숫자가 random이라는 사실을 Car가 알아야 할까요? 
Car는 랜덤 숫자, 높은 숫자, 낮은 숫자 상관없이 숫자만 받으면 조건을 따져 전진하거나 멈춰있는 친구에요.

parameter명이 조금 간결해져도 좋겠네용

sooyeoniya replied:
말씀 들어보니 도메인을 좀 더 순수하게 만들려면 `Car` 자체가 랜덤 여부를 굳이 알아야 할 필요는 없는 것 같네용. `Car`에서 `move`라는 메서드는 단순히 자동차의 움직임만 제어하는 역할이니까요 😃 도밥 말처럼 확장성을 고려해보면 파라미터에 랜덤 숫자가 들어오지 않을 수도 있겠다는 생각이 듭니다!

Creative-Lee commented on src/models/Car.js:
```diff
@@ -0,0 +1,23 @@
+  }
+
+  move(randomNumber) {
+    if (randomNumber >= RANDOM_BOUNDARY_NUMBER) this.#count++;
```
RANDOM_BOUNDARY_NUMBER라는 네이밍만 보고 직관적으로 해석이 되지 않는것 같아요.

네이밍 지옥에 빠지지는 마시고 살짝만 고민해보셔요 ㅎㅎ

Creative-Lee commented on .prettierrc:
```diff
@@ -1,3 +1,7 @@
-  "endOfLine": "auto"
+  "endOfLine": "auto",
+  "tabWidth": 2,
+  "printWidth": 200,
```
상당히 와이~~~~~~~~~~드 하군요 ㅋㅋㅋ
가독성을 이유로 printWidth는 조금씩 줄여가는걸 추천드려요~!(대충 120언더까지)

(미션에서는 바꿀필요 없어요! 작업자끼리 잘 썼다면 그만입니다 ㅎㅎ)

sooyeoniya replied:
사실 원래는 `100`으로 두었다가 변경한 부분인데, 프로그래밍 요구사항에서 `indent depth(2)` 랑 `함수 길이 제한(15)`을 맞추기 위해 저런식으로 작업했습니다 ㅎㅎ 이렇게 야매로 프로그래밍 요구사항을 해결하면 안되겠죵 😅

Creative-Lee replied:
함수 길이 제한과 인덴트는 니야가 너 나은 코드를 작성하게 유도하는 장치일 뿐이에요.
하나의 함수가 하나의 역할만 수행하고 있고, 그럼에도 길이가 길거나 depth가 깊다면 그냥 넘어가셔도 좋습니다.
>이렇게 야매로 프로그래밍 요구사항을 해결하면 안되겠죵 😅
>하지만 이런 의도였다면...ㅎ 살짝쿵 돌려놓아봅시다....🤭

woowapark replied:
@sooyeoniya 
ㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋㅋ 구경하다가 요구 사항 이야기가 눈에 띄어 코멘트 남기고 갑니다 :) 
**인덴트 요구 사항과 함수 길이 제한은 `왜` 요구 사항에 있을까요?**  
어떤 것을 고민하고 연습하기 위한 요구 사항일지 한번 생각해보고 코멘트로 남겨주시고ㅋㅋ 
생각해봤는데 모르겠다면 호이초이 손잡고 같이 찾아오세요~

Creative-Lee replied:
엇 이거 리뷰 달았었는데 누락됐네요 ㅋㅋㅋㅋ
이미 공원이 제가 드리려던 코멘트를 드렸으니 고민해보셔요~ ㅎㅎ

Creative-Lee commented on src/views/Input.js:
```diff
@@ -0,0 +1,27 @@
+  async carName() {
+    try {
+      const input = await Console.readLineAsync(INPUT_PROMPT_MESSAGE.CAR_NAMES);
+      return validateCarNames(input);
```
✔️ 
input 객체 메서드에서 validate를 하고 있어요
validate는 view로직과 domain로직 중 어디에 포함될까요?

사용자의 입력을 받는 로직과 자동자 이름을 검증하는 로직이 함께 있으면 어떤 단점이 있을까요?

sooyeoniya replied:
이 부분도 수업 때 보면서 고민했었습니다. 저는 `validate` 내용 자체는 `domain`이라고 생각했는데, `try-catch` 를 `UI` 내부에서 진행할거면 그 안에서 `validate`를 사용해 한 번에 유효성 검증된 값을 반환하자는 식으로 구현했습니다. 

확실히 이렇게 구현하게 된다면, `domain + ui` 의 결합도가 높아져서 만약 `validate` 필요 없이 입력받는 로직만 따로 필요한 경우에는 재사용성이 떨어지긴 할 것 같네요 🥹

그래서 제가 페어분께 제안드렸던 내용 중 하나가 `validate`를 외부에서 주입하고 만약 `validate` 함수가 없는 경우에는 `validate`없이 `early return` 으로 그냥 `input` 값을 반환하자였습니다.

또 다른 대안은 제가 평소에 프리코스 때 연습하면서 작성했던 방식인데, `try-catch` 문과 `validate`를 아예 `UI` 밖으로 빼내어 `controller`에서 작업하거나 `util`로 관리를 했었는데 그 방식이 더 효율적일 것 같다는 생각이 들었습니다!

이 부분에 있어서 다시 수정해볼게용!

Creative-Lee commented on __tests__/car.test.js:
```diff
@@ -0,0 +1,24 @@
+import Car from "../src/models/Car.js";
+
+describe("Car 모델 클래스 테스트", () => {
+  test.each(["Niya", "Hoyychoi"])("Car 클래스 생성 시 name 값이 잘 반환되고 count가 0으로 초기화 되는지는지 확인한다.", (name) => {
+    // given
+    // when
+    const car = new Car(name);
+    // then
+    expect(car.name).toEqual(name);
+    expect(car.count).toBe(0);
```
Car 클래스는 생성자에 전달된 인자에 대한 별다른 검증로직 없이 순수하게 초기화만 하고있어요.
그렇다면 Car클래스의 인스턴스를 만드는 기본 동작(인자로 넣은 값이 필드에 잘 초기화 되는 것)은 자바스크립트 엔진 선에서 이미 보장 된게 아닐까용?

sooyeoniya replied:
오 말씀 들어보니까 정말 그렇네요.. 초기화 하는 것 자체는 굳이 테스트 할 필요가 없군요..! 외부 라이브러리(예를 들어, 프리코스 때 사용한 Mission-utils)를 테스트 하지 않는 것처럼요.

Creative-Lee replied:
니야~! 이걸 본다면 DM 한번 봐주시겠어요? ㅋㅋㅋ

woowapark replied:
같이 고민해보시면 좋을 것 같아 여기에도 추가 코멘트만 하나 남겨둡니다 :)  

- 1 엔진에서의 초기화 여부를 검증한다는 맥락으로 본다면 불필요할 수 있을 것 같고요
- 2 도메인 규칙을 만족하는 경우의 생성 여부를 검증한다는 맥락으로 본다면 유지하는 것이 의미있을 수 있을 것 같아요 (유효한 이름이라면 자동차를 정상적으로 생성할 수 있어야 한다)

현재 class로 만들어주신 Car에서 검증 없이 바로 인스턴스를 만들기만 하다보니 1의 테스트가 된 것 같아요. 도메인과 UI를 나누고, 도메인에 대한 검증을 테스트로 해보는 연습을 하는 단계이니 이 부분은 한번 2의 맥락에서도 추가로 고민해보아도 좋을 것 같습니다 :) 
예를 들어 이 프로그램의 도메인 규칙 중에는 `자동차의 이름은 5자 이하여야 한다`는 조건이 있는데요. 이 테스트에서는 5자가 넘는 `Hoyychoi`를 넣어도 자동차가 잘 만들어지게 되네요.     

`validateCarNames` 에서 따로 검증이 되고 있긴 하지만, 만약 다른 개발자가 `validateCarNames`를 거치는 것을 깜빡하고 `new Car("여섯글자이름")` 를 해서 자동차를 만들려 시도한다면 어떻게 될까요?

Creative-Lee commented on __tests__/validateTryCount.test.js:
```diff
@@ -0,0 +1,30 @@
+    expect(() => validateTryCount(input)).toThrow(errorMessage);
+  });
+
+  test("올바른 입력 값을 받는 경우 에러가 발생하지 않는다.", () => {
```
✔️ 
팀에 새로 합류한 동료 개발자가 테스트 코드 명세만 보고 `올바른 입력 값`이 무엇인지 알 수 있을까요?
위에 실패케이스 given에 해당하는 설명처럼 자세하게 바꿔봅시다
(예외케이스 정의와도 관련있겠네요~!)

Creative-Lee reviewed (changes requested):
안녕하세요 니야 (수여'니야' 맞나요? 🤪) 자동차 경주 미션 리뷰어 도밥입니다~!
만나서 반가워요~ ㅎㅎ 우테코 크루가 되신것을 축하드립니다 🎊 
레벨 1 시작이네요! 빠이팅입니다💪 

니야의 설계 방식과 생각이 담긴 PR이라 코드를 이해하기도 훨씬 쉬웠어요. 
또 각 파일이 어떤 역할을 하는지 설명까지 첨부해주셔서 너무 좋았습니다 👍 
내용에 대해 살짝 이야기 나눠보자면

> ### ✔️ 개별 Car 클래스 생성 및 관리
> Cars 라는 클래스에서 모든 자동차의 정보를 한 번에 관리했었다.

'Cars 클래스가 모든 정보를 관리하는 구조에서 **(역할을 분리하기 위해)** Car class를 사용해  보았다' 라고 이해했는데, 아주 좋은 시도라고 생각합니다 ㅎㅎ
괄호와 같은 생각이 아니었더라도, `뭔가 이상하다` 혹은 `그냥 나눠보고싶다.`  정도의 느낌이라도 충분해요~!

>그러다보니 외부에서 전체 자동차에 대한 인스턴스를 관리해주어야 했다.

여러대의 Car를 관리하는 역할이 있는건 자연스럽게 느껴져요~!

---

> ### ✔️ Controller 분리 (GameController와 RaceController)
> 로직이 모두 하나의 Controller에서 관리되고 있었다.
> 코드가 길어지고 기능이 많아져서 응집도가 높아졌다.

작성된 코드가 `결합도`가 높아 분리의 필요성을 느낀것 같아요.
결합도가 높으면 보통, 하나의 기능을 수정하기위해 다른 기능 코드까지 수정해야하는 경우가 생겨요.

(결합도는 코드가 서로 의존 정도를, 응집도는 기능의 집중 정도라고 생각해보면 되어요)
(응집도는 높으면서 결합도는 낮을수록 이상적인 구조라고 말해요)

> GameController: 게임 자체에 대한 전체 실행 로직 관리
> RaceController: 시도횟수에 따른 각각의 경주를 실행하고 우승자를 추출하는 로직 관리

로직에 성향에 따라 class를 나눈 부분을 칭찬드리고 싶어요.
다만, 코드에선 `RaceController`가 컨트롤러의 역할은 아닌것 같아요.
경주를 하고, 각 라운드를 진행하고, 승자를 계산하는 로직은 어떤 로직인가요?
또, 컨트롤러는 어떤 로직을 가지고 있어야 하나요?
니야를 가두는 MVC라는 틀을 과감하게 벗어던지고 view와 도메인 분리에 집중해 보셔요~!

---

> ✔️ 입출력 로직 Console 객체로 관리

이 부분은 코멘트를 확인해주세요~

---

> ### ✔️클래스의 정적 메서드와 객체의 메서드 차이
> help me. 

도와드립죠. 
missonUtils의 Console 모듈은 클래스기반의 **정적메서드로만** 구현되어 있더라고요.
모든 메서드가 정적 메서드라면 class를 사용할 필요가 있을까요? 일반 객체로 관리해도 될텐데 말이죠. 그래서 저라면 일반 객체로 만들었을 것 같아요. 이게 제 기준이고요.
사실 콕 찝어 '클래스의 정적메서드와 객체 메서드의 차이' 라고하면 뭐가 없는 것 같고 취향의 차이인것 같습니다.
대신 클래스와 객체의 차이는 뭐가 많으니, 이쪽으로 한번 파고들어보셔요~!


--- 

### PR에 대해

PR이 코드 이해를 돕는 정보들로 꽉꽉 담겨있어서 너무 좋았습니다. 
앞으로도 이런식으로 정보와 생각들을 담아주시면 니야만의 무기가 될 수 있을것 같습니다.
아 그리고 지금은 개인적인 메모 느낌의 PR인데 조금 대화하는 느낌이 나면 좋을것 같네요 ㅋㅋ
(머지않아 회사동료가 읽게 될 PR이니까요~!)

너무 고생 많으셨습니다.

✔️ 마크 달려있는 코멘트만 조금 고민해보시고 수정 후 리뷰 재요청 주세요~!
(나머지 코멘트는 차근히~ step2 와 함께 반영하시는걸 추천드립니다. 별개로 코멘트에 대해 궁금한 점이나 답변은 남겨주셔요~)

sooyeoniya committed:
test: 불필요한 Car 테스트 코드 삭제

- Car 인스턴스 생성 및 초기화 하는 로직 자체를 테스트하고 있어서 불필요하고 판단하여 삭제함
- JS 내부에서 동작하는 부분이기 때문임

sooyeoniya committed:
refactor(Car): Car 객체 move 메서드의 파라미터 이름 변경

- Car 객체가 랜덤 여부를 알 필요는 없다고 판단함.

sooyeoniya committed:
refactor(validation): tryCount 최댓값 설정

- 50으로 설정.

sooyeoniya committed:
refactor(validation): 자동차가 1대인 경우 에러 처리 추가

sooyeoniya committed:
refactor(constants): 에러 메시지 수정

- 도밥 말대로 좀 더 친절하게 확실히 알 수 있도록 수정함. 에러 메시지 느낌을 통일해주었음.

sooyeoniya committed:
refactor(RaceController): 변수명에 Instance 제거

- carsInstance, carInstance, winnersInstance, winnerInstance 에서 'Instance' 모두 제거

sooyeoniya committed:
style: prettier printWidth 100으로 변경

sooyeoniya committed:
refactor(validation): 올바른 입력 테스트 케이스에 대한 테스트 명세 추가

sooyeoniya committed:
refactor(constants): 랜덤 값의 임계점 및 자동차 이름 길이 제한 변수명 수정

Creative-Lee commented on __tests__/car.test.js:
```diff
@@ -2,7 +2,7 @@ import Car from "../src/models/Car.js";
 describe("Car 모델 클래스 테스트", () => {
   test.each([
-    ["무작위 랜덤 값이 4 이상일 경우 count가 1 증가한다.", 9, 1],
+    ["무작위 랜덤 값이 4 이상일 경우 count가 1 증가한다.", 4, 1],
```
[nit] Car클래스의 메서드 테스트이니 Car가 random 숫자인지를 몰라도 된다 라는 맥락은 비슷하게 적용될것 같아요.

또한 상수를 활용해볼수도 있겠네요`move 메서드의 인자로 ${상수 명}이상의 숫자가 주어진 경우, 자동차의 count가 ${상수 명} 증가한다`.

sooyeoniya replied:
오 참고하겠습니다! 😊

Creative-Lee replied:
리뷰들을 둘러보다가 다른 관점의 리뷰도 발견해서 남겨놔봐요~! (메서드나 변수명이 변경될 때의 유연함에 대한 관점) 
- [시지프가 생각하는 테스트 명](https://github.com/woowacourse/javascript-racingcar/pull/347#discussion_r1954511718)

엇 그러고 보니 페어인 호이초이 리뷰네요 ㅋㅋㅋㅋㅋㅋ 
같은 코드 베이스로 다른 리뷰어는 어떻게 생각하는지를 보면 더 많은 학습이 되니 시간나면 쭉 돌려보셔요 ㅎㅎ
@sooyeoniya

Creative-Lee reviewed (approved):
수정 사항 반영 확인 완료했습니다 ㅎㅎ (후다닥 끝내버리셨네요 👍) 
고생 많으셨고, step1은 머지하도록 하겠습니다.
추가로 남긴 코멘트 확인해보시고 고민과 반영은 step2에서 해주시면 되겠습니다😄 

니야가 반영한 코드 리뷰를 리뷰어가 쉽게 확인할 수 있도록 코멘트에 commit 해시를 같이 남기거나 [PR 본문에 반영표를 작성](https://github.com/woowacourse/javascript-lotto/pull/192#issuecomment-1435936668)하는것도 좋습니다.
별다른 안내가 없으면, 커밋로그만으로 이전 코드와 한번 더 대조하며 봐야해서 조금 까다롭습니다..ㅎ
(물론 니야의 커밋단위가 완벽해서 수월하게 체크할 수 있었어요! 👍 )

### 예시 gif 및 캡쳐
![해시](https://github.com/user-attachments/assets/d7a46d93-478c-435d-bfed-1c2d73f20a1e)


<img width="945" alt="image" src="https://github.com/user-attachments/assets/e8d78f35-4c88-4ffa-8d46-9d73a38f70f2" />

step2에서 뵈어요👋
